lib/jobs/generate-enclosure.js

```diff
@@ -191,62 +142,75 @@ function generateEnclosureJobFactory(
+                existSign
+            );
+        })
+        .then(function(modifiedNode){
+            if (!modifiedNode){
+                return node;
+            }
+            return modifiedNode;
+        })
+        .then(function(modifiedNode){
+            var targetsItems = _.map([].concat(targets), function(targetNode) {
+                targetNode = targetNode.id || targetNode;
+                if(targetNode === node.id ) {
+                    throw new Error('Node cannot have relationship '+type+' with itself');
+                }
+                return targetNode;
+            });
 
-        return _.transform(encls, function (result, encl) {
-            var enclSn = encl.name.slice(self.enclConst.namePrefix.length, encl.name.length);
+            var index = _.findIndex(modifiedNode.relations, { relationType: type });
+            var field = ["relations.", String(index),".targets"].join("");
+            var targetsToBeAdded = {};
+            targetsToBeAdded[field] = _.uniq(targetsItems);
 
-            if (enclSn === '') {
-                // No serial number in enclosure name, jump to the next entry
-                return true;
+            // Can not make sure prevent every exception in high concurrency.
+            if (type === 'containedBy' && 
+                modifiedNode.relations[index].targets.length + targets.length > 1) {
+                throw new Error("Node "+node.id+" can only be contained by one node");
             }
 
-            if (enclSn === nodeSn) {
-                result.encl = encl;
-                // Find the match and exit the loop
-                return false;
+            // Compute node can only have one enclosure target.
+            if (type === "enclosedBy") {
+                var targetsToBeRemoveded = {};
+                targetsToBeRemoveded[field] = [modifiedNode.relations[index].targets[0]];
+                return waterline.nodes.removeListItemsByIdentifier(
+                    node.id, targetsToBeRemoveded
+                )
+                .then(function(){
+                    return targetsToBeAdded;
+                });
             }
-        }, {
-            sn: nodeSn,
-            encl: {}
-        });
-    };
-
-    /**
-     * Get target nodes of the encloses or enclosedBy relation, and
-     * remove this relation type from node's relations
-     *
-     * @param {object} node
-     * @return {array} target nodes
-     */
-    GenerateEnclosureJob.prototype._popEnclTarget = function (node) {
-        var self = this;
-        var enclRelation = _.find(node.relations, function(entry) {
-            return entry.relationType === self.enclConst.relationEncl ||
-                entry.relationType === self.enclConst.relationEnclBy;
+            return targetsToBeAdded;
+        })
+        .then(function(targetsToBeAdded){
+            return waterline.nodes.addListItemsIfNotExistByIdentifier(
+                 node.id, targetsToBeAdded
```

Comment:
without signs, is there any possibility that enclosure node will "encloses" one compute node more than once, if running this job multiple times?